napari/layers/shapes/_accelerated_triangulate.py

```diff
@@ -319,11 +342,139 @@ def _orientation(p1: np.ndarray, p2: np.ndarray, p3: np.ndarray) -> float:
+        vertices[0][1] - centroid[1], vertices[0][0] - centroid[0]
+    )
+    prev_angle = 0.0
+    for i in range(1, len(vertices)):
+        angle = (
+            atan2(vertices[i][1] - centroid[1], vertices[i][0] - centroid[0])
+            - start_angle
+        )
+        if angle < 0:
+            angle += 2 * pi
+        if angle < prev_angle:
+            return False
+        prev_angle = angle
+    return True
+
+
+@njit(cache=True, inline='always')
+def _centroid(vertices: CoordinateArray2D) -> np.ndarray:
+    """Calculate the centroid of a polygon.
+
+    Parameters
+    ----------
+    vertices : np.ndarray
+        Array of vertex coordinates with shape (N, 2)
+
+    Returns
+    -------
+    np.ndarray
+        The centroid of the polygon
+    """
+    vertices_sum = np.zeros(2, dtype=np.float32)
+    for i in range(len(vertices)):
+        vertices_sum[0] += vertices[i][0]
+        vertices_sum[1] += vertices[i][1]
+    return vertices_sum / len(vertices)
+
+
+@njit(cache=True, inline='always')
+def is_convex(vertices: CoordinateArray2D) -> bool:
+    """Check if a polygon is convex.
+
+    Parameters
+    ----------
+    vertices : np.ndarray
+        Array of vertex coordinates with shape (N, 2)
+
+    Returns
+    -------
+    bool
+        True if the polygon is convex, False otherwise
+    """
+    # A polygon is convex if all its internal angles are less than 180 degrees.
+    # This is equivalent to checking that the orientation of all consecutive
+    # triples of points is the same.
+    if len(vertices) < 3:
+        return False
+    if len(vertices) == 3:
+        return True
+    orientation_ = Orientation.collinear
+    idx = 0
+    while idx < vertices.shape[0] - 2:
+        p1 = vertices[idx]
+        p2 = vertices[idx + 1]
+        p3 = vertices[idx + 2]
+        idx += 1
+        triangle_orientation = _orientation(p1, p2, p3)
+        if triangle_orientation != Orientation.collinear:
+            orientation_ = triangle_orientation
+            break
+    if orientation_ == Orientation.collinear:
+        return False
+    while idx < vertices.shape[0] - 2:
+        p1 = vertices[idx]
+        p2 = vertices[idx + 1]
+        p3 = vertices[idx + 2]
+        idx += 1
+        triangle_orientation = _orientation(p1, p2, p3)
+        if (
+            triangle_orientation != Orientation.collinear
+            and triangle_orientation != orientation_
+        ):
+            return False
+
+    triangle_orientation = _orientation(
+        vertices[idx], vertices[idx + 1], vertices[0]
     )
-    # fmt: on
+    if (
+        triangle_orientation != Orientation.collinear
+        and triangle_orientation != orientation_
+    ):
+        return False
+
+    triangle_orientation = _orientation(
+        vertices[idx + 1], vertices[0], vertices[1]
+    )
+    if (
+        triangle_orientation != Orientation.collinear
+        and triangle_orientation != orientation_
+    ):
+        return False
```

Comment:
I can't suggest here because GH stupidly believes this code is related to the deleted line. But this is my suggestion:

```python
    # finally, check orientation when looping around to start of polygon
    n = vertices.shape[0]
    for idx in range(n - 2, n)
        p1 = vertices[idx]
        p2 = vertices[idx + 1 % n]
        p3 = vertices[idx + 2 % n]
        current_orientation = _orientation(p1, p2, p3)
        if (
            current_orientation != Orientation.collinear
            and curent_orientation != original_orientation
        ):
            # change in orientation implies non-convex
            return False
```

Reply:
What is the benefit of for loop of length 2? It's just add modulo operation?

Reply:
yes, I presumed that the reason you didn't make the complete loop is that you didn't want to run modulo always. I prefer the 2-loop to the code being duplicated, it makes it easier to know that we are just making the same check two times again.

Reply:
Maybe `for (i1, i2, i3) in [(n-2, n-1, 0), (n-1, 0, 1)]` ?